### PR TITLE
add device group management and assignment commands

### PR DESCRIPTION
**Pull request by vkhoroz**
Test script explaining new commands:

```
#$ $fioctl-local config device-group --help) 
Manage factory device groups

Usage:
  fioctl config device-group [command]

Available Commands:
  create      Create a new device groups
  delete      Delete an existing device group
  list        Show available device groups
  update      Update an existing device group

Flags:
  -h, --help   help for device-group

Global Flags:
  -c, --config string    config file (default is $HOME/.config/fioctl.yaml)
  -f, --factory string   Factory to list targets for
  -t, --token string     API token from https://app.foundries.io/settings/tokens/
  -v, --verbose          Print verbose logging

Use "fioctl config device-group [command] --help" for more information about a command.

#$ $fioctl-local device config group --help) 
Assign a device to an existing factory device group

Usage:
  fioctl devices config group <device> [<group>] [flags]

Flags:
  -h, --help    help for group
      --unset   Unset an associated device group

Global Flags:
  -c, --config string   config file (default is $HOME/.config/fioctl.yaml)
  -t, --token string    API token from https://app.foundries.io/settings/tokens/
  -v, --verbose         Print verbose logging

#$ $fioctl-local config device-group list) 
NAME      DESCRIPTION    CREATED AT
----      -----------         ----------
#$ $fioctl-local config device-group create test) 
Name: 		test
Created At: 	2020-11-24T21:16:29

#$ $fioctl-local config device-group create new) 
Name: 		new
Created At: 	2020-11-24T21:16:49

#$ $fioctl-local config device-group list) 
NAME      DESCRIPTION    CREATED AT
----      -----------         ----------
new                                   2020-11-24T21:16:49
test                                    2020-11-24T21:16:29

#$ $fioctl-local config device-group update new -n garbage) 
#$ $fioctl-local config device-group list) 
NAME      DESCRIPTION    CREATED AT
----      -----------         ----------
garbage                             2020-11-24T21:16:49
test                                    2020-11-24T21:16:29

#$ $fioctl-local config device-group delete garbage) 
#$ $fioctl-local config device-group list) 
NAME      DESCRIPTION    CREATED AT
----      -----------         ----------
test                                    2020-11-24T21:16:29

#$ $fioctl-local device show dev1) 
UUID:		43438cd1-7887-4393-80b5-45ff6ba407a8
Owner:		admin
Factory:	default
Up to date:	false
Target:		 / sha256()
Ostree Hash:	
Created:	2020-11-20T17:37:40+00:00
Last Seen:	

-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE/CCBpI27tb/B+ADP36e47ga4cKRz
au8Ty0JpjdDZNFR7aauyq+qPwrelTEDm+8DnQn0J4JQXquOMeyg+9xr9UA==
-----END PUBLIC KEY-----
#$ $fioctl-local device config group dev1 test) 
#$ $fioctl-local device show dev1) 
UUID:		43438cd1-7887-4393-80b5-45ff6ba407a8
Owner:		admin
Factory:	default
Group:		test
Up to date:	false
Target:		 / sha256()
Ostree Hash:	
Created:	2020-11-20T17:37:40+00:00
Last Seen:	

-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE/CCBpI27tb/B+ADP36e47ga4cKRz
au8Ty0JpjdDZNFR7aauyq+qPwrelTEDm+8DnQn0J4JQXquOMeyg+9xr9UA==
-----END PUBLIC KEY-----
#$ $fioctl-local config device-group delete 11) 
ERROR:  There are devices assigned to this device group
#$ $fioctl-local device config group dev1 --unset) 
#$ $fioctl-local device show dev1) 
UUID:		43438cd1-7887-4393-80b5-45ff6ba407a8
Owner:		admin
Factory:	default
Up to date:	false
Target:		 / sha256()
Ostree Hash:	
Created:	2020-11-20T17:37:40+00:00
Last Seen:	

-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE/CCBpI27tb/B+ADP36e47ga4cKRz
au8Ty0JpjdDZNFR7aauyq+qPwrelTEDm+8DnQn0J4JQXquOMeyg+9xr9UA==
-----END PUBLIC KEY-----
```

### CODE REVIEW

File: subcommands/config/device_group.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	})
+	groupCmd.AddCommand(&cobra.Command{
+		Use:   "rename <id> <new_name>",
+		Short: "Rename an existing device group",
+		Run:   doRename,
+		Args:  cobra.ExactArgs(2),
+	})
+	groupCmd.AddCommand(&cobra.Command{
+		Use:   "delete <id>",
+		Short: "Rename an existing device group",
```
Comment (mike-sul) [inline subcommands/config/device_group.go:40]: Copy-paste detected see `Rename`.

**Comment by mike-sul** [comment]
`Group:		11	test` 
Does it mean a device can be in two groups?

**Comment by mike-sul** [comment]
Maybe it makes sense to print `Group: N/A` or `Group: nil` if a device doesn't belong to any group so a user doesn't need to go through all output in order to figure out if a device belongs to some group or not.

**Comment by vkhoroz** [comment]
> 
> 
> `Group: 11 test`
> Does it mean a device can be in two groups?

It's a group ID followed by a group name, I could take a name into quotes and put a dash in between to better visualize that.

**Comment by vkhoroz** [comment]
> 
> 
> Maybe it makes sense to print `Group: N/A` or `Group: nil` if a device doesn't belong to any group so a user doesn't need to go through all output in order to figure out if a device belongs to some group or not.

I just followed the same pattern used for other fields.  Can easily change this, but want to also hear Andy's opinion.
Note that displaying nothing in this case is more backward-compatible: customers not using groups would not notice their appearance, who knows what awk scripts they use to automate their logic on top of fioctl.

File: subcommands/config/device_group.go (hunk 1 above)
Comment (vkhoroz) [reply subcommands/config/device_group.go:40]: You have an eye like a hawk, thank you!

File: client/foundries.go [hunk 2]
```diff
@@ -270,7 +292,54 @@ func NewApiClient(serverUrl string, config Config, caCertPath string) *Api {
...
+			msg := fmt.Sprintf("HTTP error during %s '%s': %s",
+				res.Request.Method, res.Request.URL.String(), res.Status)
+			if len(body) < PRINT_LIMIT {
+				// return an error response body up to a meaningful limit - if it spans beyond a few
+				// lines, need to find a more appropriate message.
+				msg = fmt.Sprintf("%s\n=%s", msg, body)
+			}
+			err = &HttpError{msg, res}
+	}
+	return &body, err
```
Comment (mike-sul) [inline client/foundries.go:339]: It looks like `body` is returned twice in the case of an error, as `body` and as msg in `err`, although I guess that's OK.

**Review by mike-sul — APPROVED** [review]

**Review by mike-sul — APPROVED** [review]

File: client/foundries.go (hunk 2 above)
Comment (vkhoroz) [reply client/foundries.go:339]: Yes, in some cases it does, and that's by design.

A caller might need a body, and if it does - it should not parse it out of an error message.

**Comment by doanac** [comment]
In general, I'd skip showing the group ids and just let the commands use the group names which is what the user care's about.

We usually do listing a little different using tabby. e.g. you might do:
~~~
#$ fioctl-local config device-group list
NAME         CREATED AT
--------     ------------------
new          2020-11-24T21:16:49
test         2020-11-24T21:16:29
~~~

**Comment by doanac** [comment]
> Maybe it makes sense to print Group: N/A or Group: nil if a device doesn't belong to any group so a user doesn't need to go through all output in order to figure out if a device belongs to some group or not.

I think other places for device-show we simply don't show the attribute if its not set.

**Comment by mike-sul** [comment]
> > Maybe it makes sense to print Group: N/A or Group: nil if a device doesn't belong to any group so a user doesn't need to go through all output in order to figure out if a device belongs to some group or not.
> 
> I think other places for device-show we simply don't show the attribute if its not set.

OK, IMHO, it's slightly less convenient, a user has to parse all output in order to figure that the value/para, they look for is null/not set.

**Comment by vkhoroz** [comment]
> In general, I'd skip showing the group ids and just let the commands use the group names which is what the user care's about.
 
I actually prefer if users operate groups by ID, not by name as it's more stable and versatile.
A name can be literally any string and might cause issues with scripting.  Not to mention that it would also result into an unnecessary database index lookup.

I've already structured all device group related APIs to work with IDs and think it's a good habit from a user perspective to use IDs, not names.

File: subcommands/config/device_group.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	groupCmd := &cobra.Command{
+		Use:   "device-group",
+		Short: "Manage factory device groups",
+	}
+	cmd.AddCommand(groupCmd)
+
+	groupCmd.AddCommand(&cobra.Command{
+		Use:   "list",
+		Short: "Show available device groups",
+		Run:   doList,
```
[the remark this answers is not in the document]
Comment (vkhoroz) [reply subcommands/config/device_group.go:24]: good point, I forgot those nasty golang name resolution rules

File: subcommands/devices/config_group.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	"os"
+	"strconv"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	groupCmd := &cobra.Command{
+		Use:   "group <device> [<id>]",
```
[the remark this answers is not in the document]
Comment (vkhoroz) [reply subcommands/devices/config_group.go:14]: I wouldn't like fioctl to do any extra calls.
If we want to support group names here - we need to support ID & name (or just name) on API side as well.

But IMHO, working with names is less robust than working with IDs for device groups.
A group can be literally any string, potentially hard to script, whereas ID is a simple number easy to use.

If you really want names here - I can find good arguments for it (like easier to remember), it simply requires more work.

**Comment by doanac** [comment]
> I actually prefer if users operate groups by ID, not by name as it's more stable and versatile.
A name can be literally any string and might cause issues with scripting. Not to mention that it would also result into an unnecessary database index lookup.

Either way you have to look it up in the DB. Nobody's going to remember that "us-east-1" is actual $some-random-number. So all fioctl interactions will become:
~~~
## look up the device group
fioctl device group list | grep
fioctl .. group-id
~~~

That's actually *harder* to script.

As per names being crazy. I think you bring up a good point. We should limit what we accept from group names. I'd say limit to values that are URL friendly. or maybe go nuts and just be `[a-z0-0_\-]*`

File: client/foundries.go [hunk 5]
```diff
@@ -378,20 +436,14 @@ func (a *Api) Post(url string, data []byte) (*[]byte, error) {
...
-	body, err := ioutil.ReadAll(res.Body)
-	if err != nil {
-		return nil, err
-	}
-
-	if res.StatusCode != 201 {
-		return nil, fmt.Errorf("Unable to POST '%s': HTTP_%d\n=%s", url, res.StatusCode, body)
-	}
-	return &body, nil
+	return readResponse(res, log)
```
[the remark this answers is not in the document]
Comment (vkhoroz) [reply client/foundries.go:446]: I was going to do that, but there is this `RawGet` thing which broke my plans.
I'm not willing to spend more time on this "generalization" in scope of unrelated work... maybe next time.

**Comment by vkhoroz** [comment]
> As per names being crazy. I think you bring up a good point. We should limit what we accept from group names. I'd say limit to values that are URL friendly. or maybe go nuts and just be `[a-z0-0_\-]*`

I'd go nuts here and only allow valid identifiers.
A thing is, I've reused the same validation rules as for factory/device names, so we probably need to go nuts for all names, next time.

File: subcommands/devices/config_group.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (doanac) [reply subcommands/devices/config_group.go:14]: you can change the API if you want. but given your passion of group-ids i thought maybe fioctl could look up the ID and then make the calls. Its 2 HTTP round trips, but the user would have made two round trips already to look it up.

**Comment by vkhoroz** [comment]
> Either way you have to look it up in the DB. Nobody's going to remember that "us-east-1" is actual $some-random-number. 
> That's actually harder to script.

That's a valid argument, and initially I was hasitant if going for IDs versus names would be better.
A cornerstone for me was that we don't have any validation for "good" names, and I didn't want to put users into a situation they pick up a "meaningful, but arbitrary" name and then have problems with scripting around it.

If we switch to use names - need more strict validation for them, change APIs to work with names, and also add an optional description field where users may add their garbage.
@doanac If you agree to all 3 points - I'll do that work tomorrow on both API and fioctl sides.

**Comment by doanac** [comment]
Sounds good. I like the idea of a description now that you mention it.

**Comment by vkhoroz** [comment]
> OK, IMHO, it's slightly less convenient, a user has to parse all output in order to figure that the value/para, they look for is null/not set.

@mike-sul I agree on convenience, but consistency is more important... and also that backward-compatibility thing.

File: subcommands/devices/config_group.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (vkhoroz) [reply subcommands/devices/config_group.go]: Changed APIs and fioctl to work with group names as per below discussion

**Comment by vkhoroz** [comment]
Continued in https://github.com/foundriesio/fioctl/pull/69 due to a force push

**Pull request closed without merging**